### ISSUE TITLE
Multi-sample analysis pipeline 

### ISSUE THREAD

**Issue by Sengels24** [prabhakarlab/Banksy_py#10]
Thanks for creating this tool. I have got a similar question as stated in #1  : is there are multi-sample analysis pipeline available for python?

Essentially, we have an adata frame containing cells from various samples (the sample origin is known for each cell). I was wondering how Banksy goes about creating a spatial graph when there are multiple samples in the same adata frame. This should be done for each sample separately (Construction of the spatial Nearest-Neighbour graph), but I am not sure how to go about it.
If I look in the slideseqv1_analysis.ipynb, match_labels is not used until cell clustering?

Thank you very much!

**Comment by chousn** [comment]
Hi Yikai, thanks for your question!

When `m > 0`, we incorporate the Azimuthal Gabor Filter (AGF) as mentioned in the paper. This filter considers the direction (angle) between a cell and its neighbors, represented as a phase shift. Imagine a compass: each neighbor's contribution depends on its location relative to the center cell. The phase shift is why the calculation involves complex numbers as an intermediate step.

You're absolutely right about the modulus. We take the absolute value (modulus) of the complex number, but only *after* summing the contributions from all neighbors with their respective phase shifts. If we took the modulus of each neighbor individually and then summed them, it would be the same as using `m = 0 `(no AGF).

The benefit of AGF is its ability to detect expression gradients. Let's say a gene is generally more highly expressed in the top half of cells compared to the bottom half. Their contributions would result in a higher overall magnitude since their phases (i.e. orientation relative to center cell) are closely matched. This will result in a higher final value compared to a scenario where the expression is evenly distributed around the cell. In simpler terms, AGF helps us identify directional trends in how genes are expressed across neighboring cells.

Regarding the number of neighbors: We use more neighbors for `m > 0` because it can improve the accuracy of capturing expression gradients. Imagine you're trying to estimate the slope of a hill. More data points (neighbors) from different directions can provide a more reliable picture.

On a separate note, we'd generally recommended to start off with m = 0 (no AGF) as it's faster and can still provide very good results. We haven't tested m > 1 extensively, but it's likely not necessary for most applications.

I hope this clarifies how the matrix works with `m > 0`!

**Comment by chousn** [comment]
Hi Sengels24, thanks for your question!

Currently, the Python version of Banksy doesn't have a built-in multi-sample analysis pipeline like the R version. However, you can achieve this functionality using anndata's concatenate function (https://anndata.readthedocs.io/en/latest/generated/anndata.AnnData.concatenate.html). Assuming all your cells are in a single dataframe, which also contains x and y coordinates as annotations:

1.⁠ **⁠Subset by Sample Origin:** Separate the cells into seperate anndata objects based on their sample origin. This ensures you analyze each sample independently.

2.⁠ **⁠Batch Correction:** Since you have multiple samples, you'll need to correct for potential batch effects. This can involve simple z-score normalization for each sample or using a tool like Harmony.

3.⁠ **⁠Compute the Neighbor Augmented Matrix (BANKSY Matrix) for Each Sample Separately:** Apply Banksy's neighbour matrix computation step to each batch-corrected sample dataset individually (this creates separate spatial graphs for each sample), generating anndata objects for each sample. 

4.⁠ ⁠**Concatenate BANKSY matrices:** Concatenate Banksy's neighbour-augmented matrices (anndata objects) along the cells dimension using ⁠ `anndata.concatenate(banksy_adata_sample1, banksy_adata_sample2, banksy_adata_sample3...) ⁠`. This merges the information from all samples into a single anndata object, which contains batch-corrected and neighbour-augmented expression features for cells from all samples.

5.⁠ **⁠Clustering on Concatenated Data:** Now you can perform clustering on the concatenated anndata object, which will generate clusters that span all the samples. 

In this approach, you wouldn't need ⁠` match_labels ⁠` because the clusters should already encompass cells from multiple samples.

We are working on an updated version of Banksy_py and will try to include an example notebook demonstrating this multi-sample workflow.

I hope this clarifies the process! Feel free to ask if you have any further questions.

**Comment by Sengels24** [comment]
Hi @chousn,

Thanks for the elaborate response! Very helpful and I will now perform analysis like this!

**Comment by Andrea-ZW** [comment]
Thank you for the clarification. I have a follow up question: once we concatenate the multi anndatasafter computed BANKSY matrices, how can we generate the clustering graph? Given their weight graphs are not generated simultaneously. And what do you mean by the clusters should already encompass cells from multiple samples.

**Comment by vipulsinghal02** [comment]
Hi Andrea and Anastasia, 

Thanks for your patience. I am just circling back here, and will try to respond to your questions by this weekend. 

Best, 
Vipul

**Comment by chousn** [comment]
Hi Anastasia,

Apologies for the slow reply, and thanks again for you patience! First, I wanted to let you know that we've recently uploaded a newer version of BANKSY (see version 1.2 branch), but this currently doesn't have a multi-sample example yet. We are working to release an example notebook (likely with the DLPFC as an example) for running multiple samples, and will keep you updated on this.

You're right that `pca_umap` does take the whole dictionary rather than individual anndata objects. The function works on the anndata object internally (referencing it via `banksy_dict[nbr_weight_decay][lambda_param]["adata"]`), and I think we will need to rewrite this function to allow it to be used more flexibly. Thanks for bringing this to our notice! For now, were you able to perform the clustering on the concatenated object?

**Comment by kousaa** [comment]
Hello,

and thanks so much for your quick response! Your tool has really saved me from a ton of extra work defining and annotating my cell subsets in my Xenium samples (which was not great with nonspatial methods) so thanks once again for building this. :-) 

I only installed your tool end of last week, so I may be already running the latest version, I will double check.

I haven't managed to perform the clustering on the concatentated data yet. Two reasons for that:

1) Let's say I generate the dict and matrix for both of my samples:

```
banksy_dict, banksy_matrix = generate_banksy_matrix(sample1, banksy_dict, lambda_list, max_m)
banksy_dict, banksy_matrix = generate_banksy_matrix(sample2, banksy_dict, lambda_list, max_m)
```

I assume I would then run this

```
banksy_dict["nonspatial"] = {
    # Here we simply append the nonspatial matrix (adata.X) to obtain the nonspatial clustering results
    0.0: {"adata": concatenate_all([sample1.X], 0, adata=sample1), }
}

banksy_dict["nonspatial"] = {
    # Here we simply append the nonspatial matrix (adata.X) to obtain the nonspatial clustering results
    0.0: {"adata": concatenate_all([sample2.X], 0, adata=sample2), }
}
```

But I am unsure how I would merge the dictionaries correctly in order to feed them to the pca_umap function. I don't work often with dictionaries so that might be more the issue here. If you have some code to showcase how to merge the two dictionaries together that would be great!

2) I also ran into an issue when trying to run the analysis with all my samples; because of the huge number of cells to be analyzed (~2 million) my jupyter kernel keeps crashing. I can run this on a server and utilize multiple cores. Is there a way to define n_jobs or run this in parallel to take advantage of these cores? Or is there a GPU option?

Any advice would be most welcome and happy to also help if I can on my end.

Cheers,
Anastasia

**Comment by chousn** [comment]
Hi Anastasia,

I've added an additional file to the banksy_1.2.1 branch: https://github.com/prabhakarlab/Banksy_py/blob/Bansky_1.2.1/banksy_utils/umap_pca_2.py

which has a function `pca_umap_adata` which should do what I mentioned earlier. Note you'll have to change the import statement since the file is named differently. I think this ought to work with either the main or the 1.2.1 branch as the dictionary structure is the same. Hope this works for now!

**Comment by kousaa** [comment]
Great! I will try to test this later on today. 

I also think that we may need a similar temporary solution for the run_Leiden_partition(), this function also seems to get a dictionary input currently.

I will play around with your suggestions about reducing memory usage. I will give you an update when I understand more why the kernel crashes, as currently there is no error message it just dies.

**Comment by kousaa** [comment]
Update on trying to use the pca_umap_adata()

<img width="1240" alt="Screenshot 2024-11-07 at 11 22 23" src="https://github.com/user-attachments/assets/b42105a4-d36c-4c07-8f59-016c1cae657e">

**Comment by chousn** [comment]
Hi Anastasia, you'll need to provide a list of pca_dims e.g. [45,] if there's only one. But I've also updated the code so it will work with a single value.

**Comment by kousaa** [comment]
This worked great, thank you! 

Any updates on the run_Leiden_partition() function? Could I maybe run this using the scanpy.tl.leiden function just by altering some of the parameter fields? Which observation is leiden expecting in order to run the clustering with the banksy generated data?

Thanks again,
Anastasia

**Comment by kousaa** [comment]
And one final question I have regarding batch effect correction, and please no rush to respond, take your time.

I have ran harmony_integrate on all my data together that generates the .obsm['X_pca_harmony'] field. Since harmony_integrate doesn't really change the gene values, how does Banksy take into account the new X_pca_harmony information?

Or would I have to use some alternative batch effect correction method (or the Z-score normalization) to apply the batch effect correction?

Thank you!
Anastasia

**Comment by chousn** [comment]
Hi Anastasia, 

Just a quick update: @lurongw is working a multi-sample notebook using four DLPFC samples as an example. This version will include per-sample normalization but does not utilize Harmony or other batch correction. We'll aim to upload it to the 1.2.1 branch by the following week. This example will illustrate how to generate a combined dictionary with different samples.

For the Harmony version, you would need to run it after the Banksy embedding step and PCA dimension reduction. Specifically, this involves performing these steps for each dataset, adding sample labels as a `.obs` column (to be used later with scanpy's `harmony_integrate` function via the `key` argument), and then concatenating the results. I believe this should also merge the PCA `.obsm` fields, but will need to check. Only then would you apply `harmony_integrate`. The final step would be to perform clustering, either using our function or scanpy's Leiden clustering. We'll work on creating a Harmony-integrated version of the notebook once the simpler version is complete!

**Comment by kousaa** [comment]
Brilliant! I am testing the code with harmony now, so I can also paste here some of the blocks of code that I am running, for you to double check that this looks legit. Feel free to use them for your tutorial if it suits the purpose.

Thank you for all your help!
Anastasia

**Comment by wlason** [comment]
Hi all! I've been luring in this thread for a while - I have a large Xenium dataset consisting of multiple patient samples which I'm analysing using scanpy/squidpy/SpatialData framework. I would love to take advantage of BANKSY clustering and compare with 'regular' UMAP/Harmony clusters - my original plan was to use the R tutorial but I feel like the python workflow will potentially be faster/more scalable, so I'd be really interested in the multi-sample tutorial when it comes out. Let me know if there is anything I can do to help @chousn and/or if you have a workflow ready for testing 😊

**Comment by chousn** [comment]
Hi Anastasia and Wojciech, @lurongw has created an example notebook with the 4 DLPFC samples (this is meant to be similar to the multi-sample example in R), which you can adapt it to run your samples. Here we're running BANKSY on each sample separately and then concatenating the result, an approach that should work on similar samples (e.g. from the same patient or same experiment). I've uploaded it to the Banksy_1.2.1 branch (*not* the main branch). Let us know if you have any issues or questions.

@lurongw is now working on a version which uses Harmony on the BANKSY embedding. Will update when ready!

**Comment by OMIC-coding** [comment]
I am eagerly anticipating the harmony version and would appreciate it if you could release it as soon as possible

**Comment by kousaa** [comment]
Hello, I have tested the code for two of my samples. I had to tweak it a bit, so it works on my end. I had all my samples in one adata file and already normalized and log-transformed, so I created the adata_list = {} dictionary and filled it in with the adata samples containing the union of hvg in the following manner:

<img width="812" alt="Screenshot 2024-11-22 at 13 42 39" src="https://github.com/user-attachments/assets/49690267-d807-4efe-b83a-d800a1ce2a8c">

And after running the run_Leiden_partition() section, I have my results_df. Is there an easy way to just plot the leiden clusters? I assume calculating ARI is not necessary right?

Thanks!
Anastasia

**Comment by kousaa** [comment]
I actually worked it out. I only needed to add the labels to my adata object.

```
# If label information is not stored in anndata object 
tmp_adata.obs['labels_' + params_name] = pd.Categorical(tmp_label.dense)
```

Thanks again,
Anastasia

**Comment by kousaa** [comment]
Hello, and please disregard if you are very busy currently. : )

I was only wondering if you have an estimate for the release of the notebook with the multisample+harmony.

Thank you! 
Anastasia

**Comment by chousn** [comment]
> One quick question I have is if I can run the `run_Leiden_partition()` function using the normal sc.pl.leiden() function. Because running it with scanpy, I can run it in parallel and really speed up the process.
> 
> Would that be ok?

Hi Anastasia, sorry for the late reply! Yes you can use the `sc.pl.leiden()` function in place of `run_Leiden_partition()` since they essentially do the same thing. However, the implementation is slightly different: although the Leiden graph clustering part is identical (basically everyone just calls the C code that the Leiden authors created), there could be significant differences in how the nearest neighbour graph is generated (note this is in expression space, not to be confused with the spatial graph that BANKSY and others use in earlier steps to define spatial neighbours) which could lead to different clustering results. Nonetheless, these are just two alternate ways of doing the clustering step so either would be ok to use.

**Comment by chousn** [comment]
> Hello, and please disregard if you are very busy currently. : )
> 
> I was only wondering if you have an estimate for the release of the notebook with the multisample+harmony.
> 
> Thank you! Anastasia

Yes we hope to try to release it within the next 2 weeks. Will keep you updated if there are any changes!

**Comment by chousn** [comment]
Hi Anastasia and Wojciech, I've uploaded a notebook that @lurongw created showing how to use Harmony on 3 DLPFC samples from different patients. This is intended to closely follow the equivalent vignette in the R package. It's in the Banksy_1.2.1 branch (not the main branch). Let us know if you have any issues or questions!

**Comment by reinertanalytics** [comment]
Hello all, thanks for a great package and especially for the integrated sample workflow! It has been really nice to use. Is there a way to get the "connections between label" plot and the colorbar showing which cluster number is which color, like for the single sample BANKSY workflow?

**Comment by lurongw** [comment]
> Hello all, thanks for a great package and especially for the integrated sample workflow! It has been really nice to use. Is there a way to get the "connections between label" plot and the colorbar showing which cluster number is which color, like for the single sample BANKSY workflow?

Hi @reinertanalytics , I have tried the following steps to subset the results_df for each sample, for plotting using plot_results. The steps below present an alternative method for plotting in the DLPFC multisample Harmony analysis. 

```
from banksy.plot_banksy import plot_results
from banksy.plot_banksy import plot_connections
from banksy.labels import Label
from banksy.cluster_methods import convert2df

target_params = ['scaled_gaussian_pc20_nc0.20_r0.40']

for counter, sample in enumerate(samples):
    for counter2, params_name in enumerate(target_params):
        tmp_df = results_df
        tmp_label = tmp_df.loc[params_name, "labels"]
        tmp_adata = tmp_df.loc[params_name, "adata"].copy()

        
        # If label information is not stored in anndata object 
        tmp_adata.obs['labels_' + params_name] = pd.Categorical(tmp_label.dense)

        # Store weights information in obsm data
        tmp_adata.obsm['weights_0'] = banksy_dict['scaled_gaussian']['weights'][0]
        
        adata_ari_temp = tmp_adata[tmp_adata.obs['sample'] == sample]
        # Subset per sample information
        weights_graph = adata_ari_temp.obsm['weights_0'][:, tmp_adata.obs['sample'] == sample]
        subset_labels = adata_ari_temp.obs['labels_' + params_name]
        subset_labels = Label(subset_labels.values.to_numpy())

        # create a new results_df for plotting
        results ={}
        results[params_name] = {
                            "decay": tmp_df.loc[params_name,'decay'],
                            "lambda_param": tmp_df.loc[params_name,'lambda_param'],
                            "num_pcs":tmp_df.loc[params_name,'num_pcs'],
                            "resolution":tmp_df.loc[params_name,'resolution'],
                            "num_labels": subset_labels.num_labels,
                            "labels": subset_labels,
                            "adata": adata_ari_temp
                        }
        results_df_subset, max_num_labels_subset = convert2df(results, True)
        print(results_df_subset)

        plot_results(
            results_df_subset,
            weights_graph,
            c_map,
            match_labels = True,
            coord_keys = coord_keys,
            max_num_labels  =  max_num_labels_subset, 
            save_path = os.path.join(file_path, 'tmp_png') ,
            save_fig = False, # save the spatial map of all clusters
            save_seperate_fig = True, # save the figure of all clusters plotted seperately
        )
```

**Comment by reinertanalytics** [comment]
Fantastic, thanks so much!

**Comment by sinkahak** [comment]
Hello! Thank you for the interesting package. I've recently transitioned from using the R version of Banksy to the Python version, primarily to leverage the NCC functionality, for a multi-sample analysis of Visium cancer datasets. My main interest lies in identifying potential differences in the TME across spatial niches, especially in relation to gene expression patterns observed in other datasets. As an initial step, I’ve applied Scanpy’s gene scoring to the neighborhood-augmented matrix.

However, I'm observing quite different results between the R and Python versions. To me it seems that the Python version tends to smooth out both sample- and region-specific differences in gene expression. Is this something that is expected? I am using the region annotations done by our lab. The R version's suggested pipeline involves:

1 running Banksy on the merged Seurat object, and
2. Harmony correction of Banksy's embedding.

Because of this, I suspect the discrepancy might be due to whether the data is merged prior to running Banksy. Would it be possible to follow a similar pipeline in Python — specifically, by staggering the spatial coordinates of different samples, and then computing the NCC graph on the merged object?

Best, Sini

**Comment by asmlgkj** [comment]
@chousn 
I noticed that the current multi-sample notebook 
https://github.com/prabhakarlab/Banksy_py/blob/Bansky_1.2.1/DLPFC_concatenate_multisample.ipynb
uses a coordinate staggering approach (concatenating samples with spatial offset), but I don't see any Harmony integration steps – neither before nor after running BANKSY. In a previous GitHub discussion, there was mention of a version that would use Harmony on the BANKSY embedding.
Specifically, I'm wondering:

What is the recommended best practice for multi-sample BANKSY analysis with batch correction?
Should the workflow be:
python# Option A: Harmony after BANKSY
- Run BANKSY on concatenated samples (with coordinate staggering)
- Apply Harmony on the BANKSY embedding

# Option B: Individual BANKSY then Harmony
- Run generate_banksy_matrix() on each sample separately  
- Concatenate the individual BANKSY matrices
- Apply Harmony integration on the combined BANKSY features

# Option C: Something else?

Could you provide guidance on the correct implementation? I believe many users would benefit from a clear multi-sample + batch correction workflow.

My Current Approach:
I'm considering processing each sample individually with generate_banksy_matrix(), then concatenating the results and applying Harmony. Would this be appropriate, or are there important considerations I'm missing?
Any insights or examples you could share would be tremendously helpful. I'd be happy to test any recommended approaches and provide feedback.
Thank you again for this fantastic tool and for making spatial transcriptomics analysis so much more accessible!

**Comment by vipulsinghal02** [comment]
@asmlgkj There are now notebooks for both multi sample and harmony in the main repo. (also, things are now pip installable).
